internal/worker/apiremotecaller/remote.go

```diff
@@ -87,17 +84,37 @@ func newRemoteServer(config RemoteServerConfig, internalStates chan string) Remo
 		apiOpener:      config.APIOpener,
 		changes:        make(chan []string),
 		internalStates: internalStates,
+		connections:    make(chan chan api.Connection),
 	}
 	w.tomb.Go(w.loop)
 	return w
 }
 
-// Connection returns the connection to the remote API server.
-func (w *remoteServer) Connection() api.Connection {
-	w.mu.Lock()
-	defer w.mu.Unlock()
+// Connection returns the current connection to the remote API server if it's
+// available, otherwise it will send the connection once it has connected
+// to the remote API server.
+// No updates about the connection will be sent to the connection channel,
+// instead it's up to the caller to watch for the broken connection and then
+// reconnect if necessary.
+func (w *remoteServer) Connection(ctx context.Context, fn func(context.Context, api.Connection) error) error {
+	ch := make(chan api.Connection, 1)
+
+	go func() {
+		select {
+		case <-w.tomb.Dying():
+		case <-ctx.Done():
+		case w.connections <- ch:
```

Comment:
I'm probably missing something, can't this be done right before we return `return w.callFunc(ctx, conn, fn)` instead of spawning this new goroutine?

Reply:
Yes, that's possible, I'll change it 😉